README.md
# django-kadabra
Django middleware for Kadabra metrics framework


CHANGES
Django-Kadabra Changelog
========================

Here you can see the full list of changes between each Flask-Kadabra release.

Version 0.1.0
-------------

First public preview release.


django_kadabra/tests/tests.py
import datetime
import kadabra

from mock import call, Mock, patch

from django.conf import settings
from django.test import SimpleTestCase
from django.test.utils import override_settings

from django_kadabra.middleware import KadabraMiddleware


NOW = datetime.datetime.utcnow()


@patch('django_kadabra.middleware.kadabra')
class MiddlewareTests(SimpleTestCase):
    def setUp(self):
        self.get_response = Mock()
        self.request = Mock()

    def get_middleware(self):
        return KadabraMiddleware(self.get_response)

    def get_metrics_mock(self, middleware):
        middleware.kadabra.metrics.return_value = Mock(
            spec=['add_count', 'set_timer', 'close']
        )
        return middleware.kadabra.metrics.return_value

    def test_ctor_kadabra_called_with_no_config(self, mock_kadabra):
        middleware = self.get_middleware()
        mock_kadabra.Kadabra.assert_called_once_with({})
        self.assertEqual(middleware.kadabra, mock_kadabra.Kadabra.return_value)

    @override_settings(KADABRA_CONFIG=Mock())
    def test_ctor_kadabra_called_with_config(self, mock_kadabra):
        middleware = self.get_middleware()
        mock_kadabra.Kadabra.assert_called_once_with(settings.KADABRA_CONFIG)
        self.assertEqual(middleware.kadabra, mock_kadabra.Kadabra.return_value)

    @patch('django_kadabra.middleware._get_now')
    def test_call_no_failure(self, mock_now, mock_kadabra):
        mock_now.side_effect = [
            NOW,
            NOW + datetime.timedelta(seconds=3)
        ]
        self.get_response.return_value.status_code = 200

        middleware = self.get_middleware()
        metrics = self.get_metrics_mock(middleware)
        response = middleware(self.request)

        self.assertEqual(response, self.get_response.return_value)
        metrics.add_count.assert_has_calls([
            call("Failure", 0),
            call("ClientError", 0),
        ])
        metrics.set_timer.assert_called_once_with(
            "RequestTime",
            NOW + datetime.timedelta(seconds=3) - NOW,
            mock_kadabra.Units.MILLISECONDS
        )
        metrics.close.assert_called_once_with()
        middleware.kadabra.send.assert_called_once_with(metrics.close.return_value)

    @patch('django_kadabra.middleware._get_now')
    def test_call_failure(self, mock_now, mock_kadabra):
        mock_now.side_effect = [
            NOW,
            NOW + datetime.timedelta(seconds=3)
        ]
        self.get_response.return_value.status_code = 500

        middleware = self.get_middleware()
        metrics = self.get_metrics_mock(middleware)
        response = middleware(self.request)

        self.assertEqual(response, self.get_response.return_value)
        metrics.add_count.assert_has_calls([
            call("Failure", 1),
            call("ClientError", 0),
        ])
        metrics.set_timer.assert_called_once_with(
            "RequestTime",
            NOW + datetime.timedelta(seconds=3) - NOW,
            mock_kadabra.Units.MILLISECONDS
        )
        metrics.close.assert_called_once_with()
        middleware.kadabra.send.assert_called_once_with(metrics.close.return_value)

    @patch('django_kadabra.middleware._get_now')
    def test_call_client_error(self, mock_now, mock_kadabra):
        mock_now.side_effect = [
            NOW,
            NOW + datetime.timedelta(seconds=3)
        ]
        self.get_response.return_value.status_code = 400

        middleware = self.get_middleware()
        metrics = self.get_metrics_mock(middleware)
        response = middleware(self.request)

        self.assertEqual(response, self.get_response.return_value)
        metrics.add_count.assert_has_calls([
            call("Failure", 0),
            call("ClientError", 1),
        ])
        metrics.set_timer.assert_called_once_with(
            "RequestTime",
            NOW + datetime.timedelta(seconds=3) - NOW,
            mock_kadabra.Units.MILLISECONDS
        )
        metrics.close.assert_called_once_with()
        middleware.kadabra.send.assert_called_once_with(metrics.close.return_value)

    @override_settings(DISABLE_KADABRA=True)
    @patch('django_kadabra.middleware._get_now')
    def test_call_metrics_not_sent_when_kadabra_disabled(self, mock_now, mock_kadabra):
        mock_now.side_effect = [
            NOW,
            NOW + datetime.timedelta(seconds=3)
        ]
        self.get_response.return_value.status_code = 400

        middleware = self.get_middleware()
        metrics = self.get_metrics_mock(middleware)
        response = middleware(self.request)

        metrics.close.assert_called_once_with()
        self.assertFalse(middleware.kadabra.send.called)



django_kadabra/tests/__init__.py
from .tests import MiddlewareTests


django_kadabra/tests/settings.py
import django
from os import path


DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': ':memory:',
    },
}

INSTALLED_APPS = ['django_kadabra']

MIDDLEWARE_CLASSES = (
    'django_kadabra.middleware.KadabraMiddleware',
)

SECRET_KEY = 'test'


django_kadabra/middleware.py
import kadabra

from django.conf import settings

from django_kadabra.utils import _get_now


class KadabraMiddleware(object):
    def __init__(self, get_response):
        self.get_response = get_response

        config = getattr(settings, 'KADABRA_CONFIG', {})
        self.kadabra = kadabra.Kadabra(config)

    def __call__(self, request):
        request.metrics = self.kadabra.metrics()
        start_time = _get_now()

        response = self.get_response(request)

        failure = 0
        client_error = 0
        if response.status_code >= 500:
            failure = 1
        elif response.status_code >= 400:
            client_error = 1

        request.metrics.set_timer(
            "RequestTime",
            _get_now() - start_time,
            kadabra.Units.MILLISECONDS
        )
        request.metrics.add_count("Failure", failure)
        request.metrics.add_count("ClientError", client_error)
        closed = request.metrics.close()
        if not getattr(settings, 'DISABLE_KADABRA', False):
            self.kadabra.send(closed)

        return response

